backend-tests/tests/test_artifact.py

```diff
@@ -0,0 +1,352 @@
+        for depend in depends:
+            args.extend(["--depends", depend])
+        for provide in provides:
+            args.extend(["--provides", provide])
+        try:
+            subprocess.call(args)
+            yield artifact
+        finally:
+            os.unlink(filename)
+            os.path.exists(artifact) and os.unlink(artifact)
+
+    def test_upload_artifact_depends_provides_valid(self, mongo, clean_mongo):
+        tenant, username, password = self.get_tenant_username_and_password()
+        auth_token = self.get_auth_token(username, password)
+
+        api_client = ApiClient(deployments.URL_MGMT)
+        api_client.headers = {}  # avoid default Content-Type: application/json
+        api_client.with_auth(auth_token)
+
+        # create and upload the mender artifact
+        with self.get_mender_artifact(
+            artifact_name="test",
+            device_types=["arm1"],
+            depends=("key1:value1", "key2:value2"),
+            provides=("key3:value3", "key4:value4", "key5:value5"),
+        ) as artifact:
+            r = api_client.with_auth(auth_token).call(
+                "POST",
+                deployments.URL_DEPLOYMENTS_ARTIFACTS,
+                files=(
+                    ("description", (None, "description")),
+                    ("size", (None, str(os.path.getsize(artifact)))),
+                    (
+                        "artifact",
+                        (artifact, open(artifact, "rb"), "application/octet-stream"),
+                    ),
+                ),
+            )
+        assert r.status_code == 201
+
+        # extract the artifact id from the Location header
+        artifact_id = r.headers.get("Location", "").rsplit("/", 1)[-1]
+        assert artifact_id != ""
+
+        # get the artifact details from the API
+        artifact_url = deployments.URL_DEPLOYMENTS_ARTIFACTS_GET.replace(
+            "{id}", artifact_id
+        )
+        r = api_client.call("GET", artifact_url)
+        assert r.status_code == 200
+        artifact = r.json()
+
+        # artifact data assertions
+        assert artifact is not None
+        assert artifact["description"] == "description"
+        assert artifact["name"] == "test"
+        assert artifact["info"] == {"format": "mender", "version": 3}
+        assert artifact["signed"] is False
+        assert len(artifact["updates"]) == 1
+        assert artifact["size"] > 0
+        assert artifact["id"] is not None
+        assert artifact["modified"] is not None
+        assert artifact["artifact_depends"] == {
+            "device_type": ["arm1"],
+            "key1": "value1",
+            "key2": "value2",
+        }
+        assert artifact["artifact_provides"] == {
+            "artifact_name": "test",
+            "key3": "value3",
+            "key4": "value4",
+            "key5": "value5",
+        }
+
+    def test_upload_artifact_depends_provides_conflicting(self, mongo, clean_mongo):
```

Comment:
Nitpick (test naming): Provides cannot conflict on upload and can (in theory) be identical for different artifacts.

Reply:
Renamed to: test_upload_artifact_provides_conflicting

Reply:
Ohh, I see my comment wasn't very clear (sorry about that). I meant something like `test_upload_artifact_depends_conflicting`.

Reply:
my fault, fixed